drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,668 @@
+        DEBUG("[sx126x] netdev: SX126X_IRQ_TIMEOUT\n");
+    }
+    else {
+        DEBUG("[sx126x] netdev: SX126X_IRQ_NONE\n");
+    }
+
+    #if IS_USED(MODULE_SX126X_STM32WL)
+        NVIC_EnableIRQ(SUBGHZ_Radio_IRQn);
+    #endif  
+}
+}
+
+                        /*RADIO HAL FUNCTIONS*/
+
+static int _write(ieee802154_dev_t *hal, const iolist_t *iolist){
+
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    size_t pos = 0;
+    /* Full buffer used for Tx */
+    sx126x_set_buffer_base_address(dev, 0x80, 0x00);
+    /* Write payload buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (iol->iol_len > 0) {
+            sx126x_write_buffer(dev, pos + 0x80, iol->iol_base, iol->iol_len);
+            DEBUG("[sx126x] netdev: send: wrote data to payload buffer.\n");
+            pos += iol->iol_len;
+        }
+    }
+
+    sx126x_set_lora_payload_length(dev, pos);
+    DEBUG("[sx126x] writing (size: %d).\n", (pos));
+    return 0;
+}
+
+static int _request_op(ieee802154_dev_t *hal, ieee802154_hal_op_t op, void *ctx){
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    (void)ctx;
+    switch (op) {
+        case IEEE802154_HAL_OP_TRANSMIT:
+        dev->pending = false;
+        ieee802154_radio_cca(hal);
+        _set_state(dev, STATE_TX);
+
+        break;
+        case IEEE802154_HAL_OP_SET_RX:
+
+        _set_state(dev, STATE_RX);
+
+        break;
+
+        case IEEE802154_HAL_OP_SET_IDLE:
+
+        _set_state(dev, STATE_IDLE);
+
+        break;
+
+        case IEEE802154_HAL_OP_CCA:
+            DEBUG("[sx126x] netdev: HAL_OP_CCA (CAD Detection state)\n");
+            dev->cad_detected = 0;
+            _set_state(dev, STATE_IDLE);
+            sx126x_set_cad(dev);
+
+        break;
+
+        default:
+        assert(false);
+        break;
+
+    }
+    return 0;
+}
+
+static int _confirm_op(ieee802154_dev_t *hal, ieee802154_hal_op_t op, void *ctx){
+    sx126x_t *dev = hal->priv;
+    (void)op;
+    (void)ctx;
+    ieee802154_tx_info_t *info = ctx;
+    bool eagain = false;
+    sx126x_state_t state;
+    _get_state(dev, &state);
+
+    
+switch (op){
+    case IEEE802154_HAL_OP_TRANSMIT:
+       if (info) {
+            info->status = (dev->cad_detected) ? TX_STATUS_MEDIUM_BUSY : TX_STATUS_SUCCESS;
+        }
+
+        
+    break;
+
+    case IEEE802154_HAL_OP_SET_RX:
+        if(state!=STATE_RX) 
+            eagain = true;
+    break;
+
+    case IEEE802154_HAL_OP_SET_IDLE:
+        eagain = (state != STATE_IDLE);
+        break;
+
+    case IEEE802154_HAL_OP_CCA:
+            *((bool*) ctx) = !dev->cad_detected;
+    break;
+
+    default:
+        eagain = false;
+        break;
+    
+}
+
+    if (eagain) {
+       
+        return -EAGAIN;
+    }
+
+    return 0;
+}
+
+static int _len(ieee802154_dev_t *hal){
+    (void)hal;
+    sx126x_t *dev = hal->priv;
+    sx126x_rx_buffer_status_t rx_buffer_status;
+    sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
```

Comment:
If the transceiver build the CRC footer (see above), this function should return the length of the packet without the 2 bytes of the CRC:
```diff
     sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
-    dev->size = rx_buffer_status.pld_len_in_bytes;
+    /* Exclude CRC */
+    dev->size = rx_buffer_status.pld_len_in_bytes - 2;
```